cpp/src/arrow/parquet/writer.cc

```diff
@@ -70,36 +72,124 @@ Status FileWriter::Impl::NewRowGroup(int64_t chunk_size) {
   return Status::OK();
 }
 
-template <typename ParquetType>
+#define CONVERT_PHYSICAL_TYPE_COPY_DATA(InType, OutType)                                \
+  template <>                                                                           \
+  Status FileWriter::Impl::ConvertPhysicalType<InType, OutType>(                        \
+      const InType* data_ptr, int64_t length, const OutType** out_ptr) {                \
+    RETURN_NOT_OK(data_buffer_.Resize(length * sizeof(OutType)));                       \
+    OutType* mutable_out_ptr = reinterpret_cast<OutType*>(data_buffer_.mutable_data()); \
+    std::copy(data_ptr, data_ptr + length, mutable_out_ptr);                            \
+    *out_ptr = mutable_out_ptr;                                                         \
+    return Status::OK();                                                                \
+  }
+
+#define CONVERT_PHYSICAL_TYPE_COPY_PTR(InType, OutType)           \
+  template <>                                                     \
+  Status FileWriter::Impl::ConvertPhysicalType<InType, OutType>(  \
+      const InType* data_ptr, int64_t, const OutType** out_ptr) { \
+    *out_ptr = reinterpret_cast<const OutType*>(data_ptr);        \
+    return Status::OK();                                          \
+  }
+
+CONVERT_PHYSICAL_TYPE_COPY_DATA(uint8_t, int32_t)
+CONVERT_PHYSICAL_TYPE_COPY_DATA(int8_t, int32_t)
+CONVERT_PHYSICAL_TYPE_COPY_DATA(uint16_t, int32_t)
+CONVERT_PHYSICAL_TYPE_COPY_DATA(int16_t, int32_t)
+CONVERT_PHYSICAL_TYPE_COPY_PTR(uint32_t, int32_t)
+CONVERT_PHYSICAL_TYPE_COPY_PTR(int32_t, int32_t)
+CONVERT_PHYSICAL_TYPE_COPY_PTR(uint64_t, int64_t)
```

Comment:
There is no difference between the two cast behaviours here, see:

```
#include <cstdint>
#include <iostream>
#include <limits>

int main() {
    uint32_t array[3] = {0, std::numeric_limits<int32_t>::max() + 1, std::numeric_limits<uint32_t>::max()};
    std::cout << array[0] << " " << array[1] << " " << array[2] << std::endl;
    int32_t* signed_array = reinterpret_cast<int32_t*>(array);
    std::cout << signed_array[0] << " " << signed_array[1] << " " << signed_array[2] << std::endl;
    std::cout << static_cast<int32_t>(array[0]) << " " << static_cast<int32_t>(array[1]) << " " << static_cast<int32_t>(array[2]) << std::endl;
}
```

Output:

```
0 2147483648 4294967295
0 -2147483648 -1
0 -2147483648 -1
```